crates/holochain/src/conductor/conductor.rs

```diff
@@ -1802,6 +1804,60 @@ mod cell_impls {
             self.running_cells
                 .share_ref(|cells| cells.keys().cloned().collect())
         }
+
+        /// Return the list of all cell IDs with DNAs "in the lineage" of the given DnaHash.
+        ///
+        /// Each DnaDef specifies a "lineage" field of DNA hashes, which indicates backward compatibility with those DNAs.
+        /// By checking a given DNA's lineage, we can determine which other DNAs are backward-compatible with it.
+        /// We also obtain a notion of forward-compatibility by checking other installed cells' DNA lineages:
+        /// if the target DNA is any installed DNA's lineage, then that DNA is forward-compatible with the target DNA.
+        ///
+        /// This function returns all installed cells which are either forward or backward compatible with the specified DNA,
+        /// including direct matches.
+        pub async fn cells_by_dna_lineage(
+            &self,
+            dna_hash: &DnaHash,
+        ) -> ConductorResult<BTreeSet<(InstalledAppId, BTreeSet<CellId>)>> {
+            // TODO: OPTIMIZE: cache the DNA lineages
+            let lineage = self.get_dna_def(dna_hash).map(|d| d.lineage);
```

Comment:
I'm finding this confusing. I believe it is saying "If you can find an earlier version of the requested app, use that"? If so, I don't think we want to do that. I think the input should be treated like a minimum version.

The way that I can see this happening is something like:
1. User installs happ A.
2. User installs happ A1, which is a successor to A.
3. User uninstalls A1, leaving it in the ribosome store.
4. User installs A again
5. user installs happ B, which depends on A1.

Now because the conductor is able to search the ribosome store, it can find that A1 was preceded by A and allow a match with B. But let's say that B, having specified A1 as a minimum version, relied on features in A1 that weren't in A (say, additional coordinator methods) - then B will not function with A as a dependency.

I think we only want to search the lineages of *other* DNAs, not the one that we were asked to search from.

Reply:
@ThetaSinner I'm not clear on your example as written, because in step 1 we install A but in step 4 we install A again without having uninstalled it.

What I am getting from this though is the claim that the dependency hash should be treated as a minimum version, which makes sense to me. I'll update that and the test.

Reply:
Made the change here https://github.com/holochain/holochain/pull/4058/commits/2de4b782422ed393774a95372fa647f5cd30caf1